src/encryption.rs

```diff
@@ -0,0 +1,357 @@
+        let mut m = HashMap::new();
+        m.insert(
+            "jwe".to_string(),
+            Box::new(JweKeyWrapper {}) as Box<dyn KeyWrapper>,
+        );
+        m
+    };
+    static ref KEY_WRAPPERS_ANNOTATIONS: HashMap<String, String> = {
+        let mut m = HashMap::new();
+        for (scheme, key_wrapper) in KEY_WRAPPERS.iter() {
+            m.insert(key_wrapper.annotation_id().to_string(), scheme.clone());
+        }
+        m
+    };
+}
+
+// EncryptLayerFinalizer can get the annotations to set for the encrypted layer
+#[derive(Debug, Default, Clone)]
+pub struct EncLayerFinalizer {
+    lbco: LayerBlockCipherOptions,
+}
+
+impl EncLayerFinalizer {
+    pub fn finalized_annotations(
+        &mut self,
+        ec: &EncryptConfig,
+        desc: &OciDescriptor,
+        finalizer: Option<&mut impl Finalizer>,
+    ) -> Result<HashMap<String, String>> {
+        let mut priv_opts = vec![];
+        let mut pub_opts = vec![];
+
+        if finalizer.is_some() {
+            finalizer.unwrap().finalized_lbco(&mut self.lbco)?;
+
+            priv_opts = serde_json::to_vec(&self.lbco.private)?;
+            pub_opts = serde_json::to_vec(&self.lbco.public)?;
+        }
+
+        let mut new_annotations = HashMap::new();
+
+        let mut keys_wrapped = false;
+        for (annotations_id, scheme) in KEY_WRAPPERS_ANNOTATIONS.iter() {
+            let key_wrapper = get_key_wrapper(scheme)?;
+
+            let mut b64_annotations = String::new();
+            if let Some(annotations) = desc.annotations.as_ref() {
+                if let Some(key_annotations) = annotations.get(annotations_id) {
+                    b64_annotations = key_annotations.clone();
+                }
+            }
+
+            b64_annotations = pre_wrap_key(key_wrapper, ec, b64_annotations, &priv_opts)?;
+            if !b64_annotations.is_empty() {
+                keys_wrapped = true;
+                new_annotations.insert(annotations_id.to_string(), b64_annotations);
+            }
+        }
+
+        if !keys_wrapped {
+            return Err(anyhow!("no wrapped keys produced by encryption"));
+        }
+
+        if new_annotations.is_empty() {
+            return Err(anyhow!("no encryptor found to handle encryption"));
+        }
+
+        new_annotations.insert(
+            "org.opencontainers.image.enc.pubopts".to_string(),
+            base64::encode(pub_opts),
+        );
+
+        Ok(new_annotations)
+    }
+}
+
+/// get_key_wrapper looks up the encryptor interface given an encryption scheme (gpg, jwe)
+pub fn get_key_wrapper(scheme: &str) -> Result<&dyn KeyWrapper> {
+    match KEY_WRAPPERS.get(scheme) {
+        Some(key_wrapper) => Ok(key_wrapper),
+        None => Err(anyhow!("key wrapper not supported!")),
+    }
+}
+
+/// get_wrapped_keys_map returns a option contains map of wrapped_keys
+/// as values and the encryption scheme(s) as the key(s)
+pub fn get_wrapped_keys_map(desc: OciDescriptor) -> Option<HashMap<String, String>> {
```

Comment:
Do we really want an `Option` here? We're always going to return a map, empty or not.

Reply:
Yes, empty map should ok, I'll fix that